File: src/main/java/rars/venus/editors/rsyntaxtextarea/RSTATokensProducer.java
package rars.venus.editors.rsyntaxtextarea;

import org.fife.ui.rsyntaxtextarea.Token;
import org.fife.ui.rsyntaxtextarea.TokenImpl;
import org.fife.ui.rsyntaxtextarea.TokenMakerBase;
import org.jetbrains.annotations.NotNull;
import rars.riscv.lang.lexing.Lexer;
import rars.riscv.lang.lexing.RVLexer;
import rars.riscv.lang.lexing.RVTokenType;
import rars.riscv.lang.lexing.TokensProducer;

import javax.swing.text.Segment;

import static rars.venus.editors.rsyntaxtextarea.RSTAUtils.tokenValue;

@SuppressWarnings("unused")
public final class RSTATokensProducer extends TokenMakerBase implements TokensProducer<Token> {
    private static final String[] LINE_COMMENT_START_AND_END = {"#", null};
    private final Lexer<Token, RSTATokensProducer> lexer;

    public RSTATokensProducer(final Lexer<Token, RSTATokensProducer> lexer) {
        this.lexer = lexer;
    }

    public RSTATokensProducer() {
        this(new RVLexer<>());
    }

    @Override
    public void addToken(
        final char[] array, final int start, final int end, @NotNull final RVTokenType tokenType,
        final int startOffset
    ) {
        this.addToken(array, start, end, tokenValue(tokenType), startOffset);
    }

    @Override
    public Token getResult() {
        return firstToken;
    }

    @Override
    public Token getTokenList(final Segment text, final int initialTokenType, final int lineOffset, final int lineNum) {
        return getTokenList(text, initialTokenType, lineOffset);
    }

    @Override
    public Token getTokenList(final Segment segment, final int initialTokenType, final int initialOffset) {
        resetTokenList();
        return lexer.getTokensList(segment, initialTokenType, initialOffset, this);
    }

    @Override
    public @NotNull Token getEmptyResult() {
        return new TokenImpl();
    }

    @Override
    public String[] getLineCommentStartAndEnd(final int languageIndex) {
        return LINE_COMMENT_START_AND_END;
    }
}

File: src/main/java/rars/venus/editors/rsyntaxtextarea/RSTATokensProducer.kt
package rars.venus.editors.rsyntaxtextarea

import org.fife.ui.rsyntaxtextarea.Token
import org.fife.ui.rsyntaxtextarea.TokenImpl
import org.fife.ui.rsyntaxtextarea.TokenMakerBase
import rars.riscv.lang.lexing.LexerOld
import rars.riscv.lang.lexing.RVLexer
import rars.riscv.lang.lexing.RVTokenType
import rars.riscv.lang.lexing.TokensProducer
import rars.venus.editors.rsyntaxtextarea.RSTAUtils.tokenValue
import javax.swing.text.Segment

@Suppress("unused")
class RSTATokensProducer @JvmOverloads constructor(
    private val lexer: LexerOld<Token, RSTATokensProducer> = RVLexer()
) : TokenMakerBase(), TokensProducer<Token> {
    override fun addToken(
        array: CharArray, start: Int, end: Int, tokenType: RVTokenType,
        startOffset: Int
    ) {
        addToken(array, start, end, tokenType.tokenValue, startOffset)
    }

    override fun getResult(): Token = firstToken

    override fun getTokenList(
        text: CharSequence,
        initialTokenType: Int,
        lineOffset: Int,
        lineNum: Int
    ): Token = getTokenList(text as Segment, initialTokenType, lineOffset)

    override fun getTokenList(
        segment: Segment,
        initialTokenType: Int,
        initialOffset: Int
    ): Token {
        resetTokenList()
        return lexer.getTokensList(
            segment,
            initialTokenType,
            initialOffset,
            this
        )
    }

    override fun getEmptyResult(): Token = TokenImpl()

    override fun getLineCommentStartAndEnd(languageIndex: Int): Array<String?> =
        LINE_COMMENT_START_AND_END

    companion object {
        private val LINE_COMMENT_START_AND_END: Array<String?> =
            arrayOf("#", null)
    }
}
